pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -305,10 +382,48 @@ protected void handleOutOfOrderEvent(Object currentComparisonValue, Object recor
     }
   }
 
+  protected void startOperation() {
+    _numPendingOperations.getAndIncrement();
+  }
+
+  protected void finishOperation() {
+    if (_numPendingOperations.decrementAndGet() == 0) {
+      synchronized (_numPendingOperations) {
+        _numPendingOperations.notifyAll();
+      }
+    }
+  }
+
+  @Override
+  public void stop() {
+    _stopped = true;
+    int numPendingOperations = _numPendingOperations.decrementAndGet();
```

Comment:
It should be just `_numPendingOperations.get()` right? No need to decrement here.

Reply:
Ohh is it because the AtomicInteger is initialized as 1 and not 0.